File: leetcode/15_3Sum.py
import unittest
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        nums.sort()
        i = 0
        while i < n - 2 and nums[i] <= 0:
            if i == 0 or nums[i] > nums[i - 1]:
                j, k = i + 1, n - 1
                while j < k and nums[i] + nums[j] <= 0:
                    if j > i + 1 and nums[j] == nums[j - 1]:
                        j += 1
                    elif k < n - 1 and nums[k] == nums[k + 1]:
                        k -= 1 
                    else:
                        val = nums[i] + nums[j] + nums [k]
                        if val == 0:
                            res.append([nums[i], nums[j], nums[k]])
                            j += 1
                        elif val > 0:
                            k -= 1
                        else:
                            j += 1
            i += 1
        return res

class Test3Sum(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().threeSum([-1,0,1,2,-1,-4]), [[-1,-1,2],[-1,0,1]])    

    def test_2(self):
        self.assertEqual(Solution().threeSum([]), [])    

    def test_3(self):
        self.assertEqual(Solution().threeSum([0]), [])    

File: leetcode/15_3Sum.rs
struct Solution { }

impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums =  nums;
        nums.sort_unstable();
        let mut res = vec![];
        let n = nums.len();
        let mut i: usize = 0;
        while i + 2 < n && nums[i] <= 0 {
            if i == 0 || nums[i] != nums[i - 1] {
                let (mut j, mut k) = (i + 1, n - 1);
                while j < k && nums[i] + nums[j] <= 0 && nums[k] >= 0 {
                    if j > i + 1 && nums[j] == nums[j - 1] {
                        j += 1;
                    }else if k < n - 1 && nums[k] == nums[k + 1]{
                        k -= 1;
                    }else{
                        let val = nums[i] + nums[j] + nums[k];
                        if val == 0 {
                            res.push(vec![nums[i], nums[j], nums[k]]);
                            j += 1;
                            k -= 1;
                        } else if val > 0 {
                            k -= 1;
                        } else {
                            j += 1;
                        }
                    }
                }    
            }
            i += 1;
        }
        res
    }
}

#[cfg(test)]
mod three_sum_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::three_sum(vec![-1,0,1,2,-1,-4]), vec![vec![-1,-1,2], vec![-1,0,1]]);
    }

    #[test]
    fn test_2() {
        let empty: Vec<Vec<i32>> = vec![]; 
        assert_eq!(Solution::three_sum(vec![]), empty);
    }

    #[test]
    fn test_3() {
        let empty: Vec<Vec<i32>> = vec![]; 
        assert_eq!(Solution::three_sum(vec![0]), empty);
    }

    #[test]
    fn test_4() {
        assert_eq!(Solution::three_sum(vec![0,0,0]), vec![[0,0,0]]);
    }

}